### DISCUSSION TITLE
Implementing a wider selection of variable transformations

### DISCUSSION THREAD

**Discussion by iraur** [pymc-labs/pymc-marketing#442] [Q&A] [answered]
Hello!
I was wondering if anyone has considered implementing alternatives for variable transformations, namely Weibull PDF for adstock where the (cumulative) decay rates are given by :
$$G_{k,\lambda}(t) = \frac{k}{\lambda}\Big(\frac{t}{\lambda} \Big)^{k-1}e^{-(\frac{t}{\lambda})^k}$$
and the Hill function for saturation, where transformed spend in week $t$ is given by :
$$x_t^{\textrm{transf}} = \frac{x_t^\alpha}{x_t^\alpha + \gamma^\alpha}$$
The main motivations for this are :
- Weibull PDF offers a lot more flexibility, compared to geometric adstock which is sometimes criticised for being too simplistic. As a result it can lead to better performing models (see [this white paper](https://ekimetrics.com/wp-content/uploads/2020/05/Ekimetrics_Facebook_White-paper.pdf) from Ekimetrics for example)
- Weibull PDF is preferred by business stakeholders as it can better model real life marketing effects
- We started out our MMM journey using [Robyn](https://facebookexperimental.github.io/Robyn/docs/welcome), Meta’s open-source MMM package, which uses Weibull PDF and Hill - it would be great for pymc-marketing and Robyn to align on these transformations (more direct comparison, QA checks, etc)

We have attempted to write a scrappy first version of these functions below and would very much appreciate any comments or feedback (still getting up to speed on pymc so appreciate your patience on any silly mistakes here!) 🤗
```
import pymc as pm
from pymc_marketing.mmm.transformers import batched_convolution

def weibull_pdf_adstock(x, shape, scale, l_max: int = 30, axis: int = 0):
    shape = pm.floatX(pt.as_tensor(shape)[..., None])
    scale = pm.floatX(pt.as_tensor(scale)[..., None])

    w = shape * pt.power(scale, -shape) * pt.power(pt.arange(l_max, dtype=x.dtype) + 1, shape - 1) * pt.power(pt.as_tensor(math.exp(1)), -pt.power((pt.arange(l_max, dtype=x.dtype) + 1)/scale, shape))
    w = (w - pt.min(w))/(pt.max(w) - pt.min(w))
    return batched_convolution(x, w, axis=axis)

def saturation_hill(x, alpha, gamma, x_marginal=None):
    inflexion = np.dot(np.array([1 - gamma, gamma]), np.array([pt.min(x, axis=0), pt.max(x, axis=0)])) # linear interpolation by dot product
    
    if x_marginal is None:
        x_scurve = x**alpha / (x**alpha + inflexion**alpha)
    else:
        x_scurve = x_marginal**alpha / (x_marginal**alpha + inflexion**alpha)
    return x_scurve
```
Thanks! cc. @louismagowan

**Comment by williambdean** [comment]
I would love to see this as part of the mix! 😃

**Comment by twiecki** [comment]
@iraur Thanks for posting. I think the best way forward would be to open a PR, it's fine if the implementation is scrappy, just from there it's easier to provide specific feedback.

**Comment by williambdean** [comment, marked as answer]
This PR #632 is in the works and will allow for users to use various saturation and adstock transformation. The order of the transformation will also be exchangeable. 

User defined transformations will also be supported!

These will all work with our lift tests integration as well

If you have feedback while the PR is still open, feel free to chime in
